### PR TITLE
[CIT-355] Refactor thin edge json module to use Group trait

### PR DESCRIPTION
**Pull request by makr11st**
## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs [hunk 1]
```diff
@@ -55,6 +59,12 @@ impl C8yJsonSerializer {
             return Err(MeasurementStreamError::UnexpectedEndOfData.into());
         }
 
+        if !self.timestamp_present {
+            self.timestamp(self.default_timestamp)?;
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_translator_lib/src/serializer.rs:63]: I feel the parser should add the timestamp, @Didier can clarify this.

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs [hunk 2]
```diff
@@ -55,6 +59,12 @@ impl C8yJsonSerializer {
...
+        }
+
+        assert!(self.timestamp_present);
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_translator_lib/src/serializer.rs:66]: Whats the need of this assert here?

File: mapper/thin_edge_json/src/group.rs [hunk 3]
```diff
@@ -105,17 +105,23 @@ mod tests {
     use mockall::predicate::*;
     use mockall::*;
 
+    #[derive(thiserror::Error, Debug, Clone)]
+    pub enum TestError {
+        #[error("test")]
+        _Test,
```
Comment (PradeepKiruvale) [inline mapper/thin_edge_json/src/group.rs:111]: is it the complete error message or something is missing?

File: mapper/thin_edge_json/src/json.rs [hunk 4]
```diff
@@ -1,165 +1,250 @@
...
     }
 
-    pub fn from_str(
-        json_string: &str,
-        timestamp: DateTime<FixedOffset>,
-    ) -> Result<ThinEdgeJson, ThinEdgeJsonError> {
-        match json::parse(&json_string) {
-            Ok(thin_edge_obj) => ThinEdgeJson::from_json(thin_edge_obj, timestamp),
-            Err(err) => Err(ThinEdgeJsonError::new_invalid_json(json_string, err)),
+    fn finish(self) -> Result<ThinEdgeJson, ThinEdgeJsonError> {
```
Comment (PradeepKiruvale) [inline mapper/thin_edge_json/src/json.rs:61]: end/build would be better name I feel.

File: mapper/thin_edge_json/src/json.rs [hunk 5]
```diff
@@ -1,165 +1,250 @@
...
-                value,
-            };
-            Ok(single_value)
-        } else {
-            Err(ThinEdgeJsonError::InvalidThinEdgeJsonNumber { name: name.into() })
+fn parse_from_rfc3339(timestamp: &str) -> Result<DateTime<FixedOffset>, ThinEdgeJsonError> {
+    let time = DateTime::parse_from_rfc3339(&timestamp).map_err(|err| {
+        ThinEdgeJsonError::InvalidTimestamp {
+            value: String::from(timestamp),
+            from: err,
```
Comment (PradeepKiruvale) [inline mapper/thin_edge_json/src/json.rs:227]: Nothing has changed in this function, why its showing as while function  has changed?

File: mapper/thin_edge_json/src/measurement.rs [hunk 6]
```diff
@@ -151,7 +151,7 @@ pub trait FlatMeasurementVisitor {
 /// ```
 pub trait GroupedMeasurementVisitor {
     /// Error type specific to this way of collecting measurements
-    type Error;
+    type Error: std::error::Error + std::fmt::Debug;
```
Comment (PradeepKiruvale) [inline mapper/thin_edge_json/src/measurement.rs:154]: does this mean Error could be either sted::error/ std::fmt:Debug?.

**Review by PradeepKiruvale — COMMENTED** [review]
Left couple of comments

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs (hunk 1 above)
Comment (didier) [reply mapper/cumulocity/c8y_translator_lib/src/serializer.rs:63]: @PradeepKiruvale Hi, I think you meant @didier-wenzek.

Comment (PradeepKiruvale) [reply mapper/cumulocity/c8y_translator_lib/src/serializer.rs:63]: Hmm, exactly, sorry.

File: mapper/thin_edge_json/src/measurement.rs (hunk 6 above)
Comment (didier-wenzek) [reply mapper/thin_edge_json/src/measurement.rs:154]: This type specification should also be added to the `FlatMeasurementVisitor` associated `Error` type.

File: mapper/thin_edge_json/src/json.rs [hunk 7]
```diff
@@ -1,165 +1,250 @@
...
-//! It also serialize the ThinEdgeJson data.
-
-use chrono::format::ParseError;
-use chrono::prelude::*;
+use crate::measurement::GroupedMeasurementVisitor;
+use chrono::{format::ParseError, prelude::*};
 use json::JsonValue;
+
 /// ThinEdgeJson is represented in this struct
 /// Since json does not understand DateTime format, the time stamp is represented as a string
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/json.rs:6]: This comment is outdated.

File: mapper/thin_edge_json/src/json.rs [hunk 8]
```diff
@@ -1,165 +1,250 @@
...
+                    self.measurements.push(ThinEdgeValue::Multi(group))
+                }
+            }
+            None => return Err(ThinEdgeJsonError::UnexpectedEndOfGroup),
+        }
+        Ok(())
+    }
+}
+
+pub struct ThinEdgeJsonParser;
```
Comment (didier-wenzek) [inline mapper/thin_edge_json/src/json.rs:125]: This struct seems useless. Why not just a function `parse_thin_edge_json(input: &[u8], visitor: &mut T)` ?

**Review by didier-wenzek — COMMENTED** [review]
The overall direction is correct. However, I would consider to use a `MeasurementGrouper` struct instead of the former `ThinEdgeJson ` struct as the intermediate struct between the json object returned by serde and the accepted visitor. Both approaches have different tradeoffs and it would be good to compare. Not necessarily in this PR, though.

File: mapper/thin_edge_json/src/json.rs (hunk 5 above)
Comment (makr11st) [reply mapper/thin_edge_json/src/json.rs:227]: It has moved due to addition of new functions.

File: mapper/cumulocity/c8y_translator_lib/examples/multi_value_translation.rs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+      "temperature": 0 ,
+      "location": {
+        "latitude": 32.54,
+        "longitude": -117.67,
+        "altitude": 98.6
+      },
+      "pressure": 98
+   }"#;
+
+    println!("\nThin_Edge_Json: \n{:#}", multi_value_thin_edge_json);
```
Comment (mneumann) [inline mapper/cumulocity/c8y_translator_lib/examples/multi_value_translation.rs:14]: drop leading "\n".

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 10]
```diff
@@ -1,113 +1,82 @@
...
-//! use c8y_translator_lib::json::CumulocityJson;
+//! use c8y_translator_lib::json::from_thin_edge_json;
 //! let single_value_thin_edge_json = r#"{
 //!        "time": "2020-06-22T17:03:14.000+02:00",
 //!        "temperature": 23,
 //!        "pressure": 220
 //!     }"#;
-//! let output = CumulocityJson::from_thin_edge_json(
-//!             &String::from(single_value_thin_edge_json).into_bytes());
+//! let output = from_thin_edge_json(single_value_thin_edge_json.as_bytes());
```
Comment (mneumann) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:13]: you can drop `.as_bytes()` when you use a raw byte literal: `br#"{....}"#`

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 11]
```diff
@@ -1,113 +1,82 @@
...
 use crate::serializer;
 use chrono::prelude::*;
-use thin_edge_json::{
-    json::{ThinEdgeJson, ThinEdgeValue},
-    measurement::GroupedMeasurementVisitor,
-};
-
-#[derive(Debug, Eq, PartialEq)]
-pub struct CumulocityJson;
+use thin_edge_json::json::{ThinEdgeJsonParserError, *};
```
Comment (mneumann) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:18]: Just use "*". No need for explicit ThinEdgeJsonParserError import.

**Review by mneumann — APPROVED** [review]
Looks good. I had some minor points... see comments.

File: mapper/cumulocity/c8y_translator_lib/src/json.rs (hunk 11 above)
Comment (makr11st) [reply mapper/cumulocity/c8y_translator_lib/src/json.rs:18]: Thanks for catching it.

File: mapper/cumulocity/c8y_translator_lib/examples/translate_to_c8yjson.rs [hunk 12]
```diff
@@ -53,12 +11,10 @@ fn thin_edge_translation_with_type_and_time_stamp() {
...
         single_value_thin_edge_json_with_type_and_time
     );
-    let output = from_thin_edge_json(
-        &String::from(single_value_thin_edge_json_with_type_and_time).into_bytes(),
-    );
+    let output = from_thin_edge_json(single_value_thin_edge_json_with_type_and_time.as_bytes());
     match output {
         Ok(vec) => {
-            println!("{}", String::from_utf8(vec).unwrap());
+            println!("{}", vec);
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_translator_lib/examples/translate_to_c8yjson.rs]: Printing just the vec is not user-friendly. User cant interpret the u8 bytes

File: mapper/thin_edge_json/src/json.rs [hunk 13]
```diff
@@ -249,63 +251,66 @@ fn input_prefix(input: &str, len: usize) -> String {
 
 #[derive(thiserror::Error, Debug)]
 pub enum ThinEdgeJsonError {
-    #[error("Invalid UTF8: {from}: {input_excerpt}...")]
-    InvalidUtf8 {
-        input_excerpt: String,
-        from: std::str::Utf8Error,
-    },
+    #[error("... time stamp within a group")]
```
Comment (PradeepKiruvale) [inline mapper/thin_edge_json/src/json.rs:254]: dint get what is the significance of the leading dots here

File: mapper/cumulocity/c8y_translator_lib/examples/multi_value_translation.rs [hunk 14]
```diff
@@ -15,7 +15,7 @@ fn multi_value_translation() {
     let output = from_thin_edge_json(multi_value_thin_edge_json.as_bytes());
     match output {
         Ok(vec) => {
-            println!("{}", String::from_utf8(vec).unwrap());
+            println!("{:?}", vec);
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_translator_lib/examples/multi_value_translation.rs:18]: will it print in user-readable format?

File: mapper/cumulocity/c8y_translator_lib/examples/single_value_translation.rs [hunk 15]
```diff
@@ -11,7 +11,7 @@ fn single_value_translation() {
     let output = from_thin_edge_json(single_value_thin_edge_json.as_bytes());
     match output {
         Ok(vec) => {
-            println!("{}", String::from_utf8(vec).unwrap());
+            println!("{:?}", vec);
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_translator_lib/examples/single_value_translation.rs:14]: Same as above

File: mapper/cumulocity/c8y_translator_lib/examples/translate_to_c8yjson.rs [hunk 16]
```diff
@@ -14,7 +14,7 @@ fn thin_edge_translation_with_type_and_timestamp() {
     let output = from_thin_edge_json(single_value_thin_edge_json_with_type_and_time.as_bytes());
     match output {
         Ok(vec) => {
-            println!("{}", vec);
+            println!("{:?}", vec);
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_translator_lib/examples/translate_to_c8yjson.rs:17]: Same as above

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 17]
```diff
@@ -216,7 +216,7 @@ mod tests {
                    }}
                 }}"#, time, measurement, measurement);
 
-        let output = from_thin_edge_json(&input.into_bytes()).unwrap();
+        let output = from_thin_edge_json(input.as_bytes()).unwrap();
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:219]: unwraps can be removed in the tests as well.

**Review by PradeepKiruvale — COMMENTED** [review]
LGTM, left a couple of comments.

**Review by didier-wenzek — COMMENTED** [review]
The current version of this PR doesn't compile with rustc 1.47.0

* It's okay with rustc 1.50.0.
* It's okay with rustc 1.47.0, *provided* an additional type constraint to the `MeasurementGrouper::accept` method.

The fixed signature is:

```
    pub fn accept<V, E>(&self, visitor: &mut V) -> Result<(), E>
    where
        V: GroupedMeasurementVisitor<Error = E>,
        E: std::error::Error + std::fmt::Debug,
    {
        // ...
    }
```

**Comment by makr11st** [comment]
> The current version of this PR doesn't compile with rustc 1.47.0
> 
> * It's okay with rustc 1.50.0.
> * It's okay with rustc 1.47.0, _provided_ an additional type constraint to the `MeasurementGrouper::accept` method.
> 
> The fixed signature is:
> 
> ```
>     pub fn accept<V, E>(&self, visitor: &mut V) -> Result<(), E>
>     where
>         V: GroupedMeasurementVisitor<Error = E>,
>         E: std::error::Error + std::fmt::Debug,
>     {
>         // ...
>     }
> ```

Thanks for catching this. I'll add the fix asap.

**Review by didier-wenzek — APPROVED** [review]
Approved.

**Pull request merged**